src/memory/memoryManager.ts

```diff
@@ -32,6 +32,14 @@ export class MemorySegmentManager {
         return insertResult;
       }
     }
-    return address.add(UnsignedInteger.toUint64(BigInt(data.length)));
+
+    const segmentSize = UnsignedInteger.toUint64(BigInt(data.length));
+    this.segmentSizes.set(address.getSegmentIndex(), segmentSize);
+
+    return address.add(segmentSize);
+  }
+
+  getSegmentSize(segmentIndex: Uint64): Uint64 {
+    return this.segmentSizes.get(segmentIndex) || UnsignedInteger.toUint64(0n);
```

Comment:
i recommend operator `??` rather than || 

```
const a = null;
const b = undefined;
const c = "";
const d = 0;
const e = false;

console.log(a || "default");  // "default"
console.log(b || "default");  // "default"
console.log(c || "default");  // "default"
console.log(d || "default");  // "default"
console.log(e || "default");  // "default"

console.log(a ?? "default");  // "default"
console.log(b ?? "default");  // "default"
console.log(c ?? "default");  // ""
console.log(d ?? "default");  // 0
console.log(e ?? "default");  // false
```